include/kernels/reduce_core.cuh

```diff
@@ -642,66 +624,27 @@ namespace quda
         w = tmpy;
         norm2(y);
     */
-    template <typename ReduceType, typename Float2, typename FloatN>
-    struct quadrupleCG3UpdateNorm_ : public ReduceFunctor<ReduceType, Float2, FloatN> {
-      Float2 a, b;
-      quadrupleCG3UpdateNorm_(const Float2 &a, const Float2 &b) : a(a), b(b) { ; }
-      FloatN tmpx {}, tmpy {};
-      __device__ __host__ void operator()(ReduceType &sum, FloatN &x, FloatN &y, FloatN &z, FloatN &w, FloatN &v)
-      {
-        tmpx = x;
-        tmpy = y;
-        x = b.x * (x + a.x * y) + b.y * z;
-        y = b.x * (y - a.x * v) + b.y * w;
-        z = tmpx;
-        w = tmpy;
-        norm2_<ReduceType>(sum, y);
-      }
-      static int streams() { return 7; } //! total number of input and output streams
-      static int flops() { return 16; }  //! flops per element check if it's right
-    };
-
-    /**
-       void doubleCG3InitNorm(d a, V x, V y, V z){}
-        y = x;
-        x -= a*z;
-        norm2(x);
-    */
-    template <typename ReduceType, typename Float2, typename FloatN>
-    struct doubleCG3InitNorm_ : public ReduceFunctor<ReduceType, Float2, FloatN> {
-      Float2 a;
-      doubleCG3InitNorm_(const Float2 &a, const Float2 &b) : a(a) { ; }
-      __device__ __host__ void operator()(ReduceType &sum, FloatN &x, FloatN &y, FloatN &z, FloatN &w, FloatN &v)
+    template <typename reduce_t, typename real>
+    struct quadrupleCG3UpdateNorm_ : public ReduceFunctor<reduce_t> {
+      static constexpr write<1, 1, 1, 1> write{ };
+      const real a;
+      const real b;
+      quadrupleCG3UpdateNorm_(const real &a, const real &b) : a(a), b(b) { ; }
+      template <typename T> __device__ __host__ void operator()(reduce_t &sum, T &x, T &y, T &z, T &w, T &v)
       {
-        y = x;
-        x -= a.x * z;
-        norm2_<ReduceType>(sum, x);
-      }
-      static int streams() { return 3; } //! total number of input and output streams
-      static int flops() { return 5; }   //! flops per element
-    };
-
-    /**
-       void doubleCG3UpdateNorm(d a, d b, V x, V y, V z){}
-        tmp = x;
-        x = b*(x-a*z) + (1-b)*y;
-        y = tmp;
-        norm2(x);
-    */
-    template <typename ReduceType, typename Float2, typename FloatN>
-    struct doubleCG3UpdateNorm_ : public ReduceFunctor<ReduceType, Float2, FloatN> {
-      Float2 a, b;
-      doubleCG3UpdateNorm_(const Float2 &a, const Float2 &b) : a(a), b(b) { ; }
-      FloatN tmp {};
-      __device__ __host__ void operator()(ReduceType &sum, FloatN &x, FloatN &y, FloatN &z, FloatN &w, FloatN &v)
-      {
-        tmp = x;
-        x = b.x * (x - a.x * z) + b.y * y;
-        y = tmp;
-        norm2_<ReduceType>(sum, x);
+#pragma unroll
+        for (int i = 0; i < x.size(); i++) {
+          auto tmpx = x[i];
+          auto tmpy = y[i];
+          x[i] = b * (x[i] + a * y[i]) + ((real)1.0 - b) * z[i];
```

Comment:
Do you have a feeling for why the bugfix of replacing `b` with `1-b` (based on the wasn't caught before?

Reply:
Previously, `b` was a two-component number with `b.y = 1 - b.x`.  This replacement with a real-valued `b` is equivalent and in my option cleaner.